include/cuco/detail/static_map/static_map_ref.inl

```diff
@@ -721,16 +821,44 @@ class operator_impl<
     }
   }
 
-  template <typename Value>
-  __device__ void insert_or_apply(cooperative_groups::thread_block_tile<cg_size> const& group,
-                                  Value const& value,
-                                  cuco::op::reduce::sum_tag)
+  template <bool PayloadWrite, typename Value, typename Op>
+  [[nodiscard]] __device__ insert_result attempt_insert_or_apply(value_type* address,
+                                                                 value_type const& expected,
+                                                                 Value const& desired,
+                                                                 Op op) noexcept
   {
-    auto& ref_ = static_cast<ref_type&>(*this);
-    ref_.insert_or_apply(
-      group, value, [](cuda::atomic_ref<T, Scope> payload_ref, T const& payload) {
-        payload_ref.fetch_add(payload, cuda::memory_order_relaxed);
-      });
+    ref_type& ref_ = static_cast<ref_type&>(*this);
+
+    if constexpr (sizeof(value_type) <= 8) {
+      return ref_.impl_.packed_cas(address, expected, desired);  // no need to wait for payload
+    } else {
+      using mapped_type = T;
+
+      cuda::atomic_ref<key_type, Scope> key_ref(address->first);
+      auto expected_key  = expected.first;
+      auto const success = key_ref.compare_exchange_strong(
+        expected_key, static_cast<key_type>(desired.first), cuda::memory_order_relaxed);
+
+      // if key success
+      if (success) {
+        cuda::atomic_ref<mapped_type, Scope> payload_ref(address->second);
+        if constexpr (PayloadWrite) {
```

Comment:
When `PayloadWrite` is true, it seems never to "apply".

Reply:
Yes. PayloadWrite, actually writes the payload using **atomic_store**. PayloadWrite is **true** only if `init_value != sentinel` **and** `value_type size > 8`